client/my-sites/hosting/sftp-card/index.js

```diff
@@ -0,0 +1,101 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { get, map } from 'lodash';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import Card from 'components/card';
+import CardHeading from 'components/card-heading';
+import MaterialIcon from 'components/material-icon';
+import Button from 'components/button';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import {
+	requestAtomicSFTPDetails,
+	resetAtomicSFTPUserPassword,
+	createAtomicSFTPUser,
+} from 'state/data-getters';
+
+const SFTPCard = ( { translate, username, password, errorCode, siteId, loading } ) => {
+	const sftpData = {
+		[ translate( 'URL' ) ]: 'sftp1.wordpress.com',
+		[ translate( 'Port' ) ]: 22,
+		[ translate( 'Username' ) ]: username,
+	};
+
+	return (
+		<Card>
+			<div className="sftp-card__icon-col">
+				<MaterialIcon icon="cloud" size={ 32 } />
+			</div>
+			<div>
+				<CardHeading>{ translate( 'SFTP Information' ) }</CardHeading>
+				<p>{ translate( "Access and edit your website's files directly using an FTP client." ) }</p>
+				{ password && (
+					<div className="sftp-card__callout-box">
+						<p>{ translate( 'Your new password for your sftp user is:' ) }</p>
+						<p>
+							<code>{ password }</code>
+						</p>
+						<strong>
+							{ translate(
+								'Make sure to save this password in a safe place! You will need to reset this password if you lose it.'
+							) }
+						</strong>
+					</div>
+				) }
+				{ errorCode === 404 && (
+					<Button onClick={ createAtomicSFTPUser } primary>
+						Create SFTP User
+					</Button>
+				) }
+				<table
+					className={ classNames( 'sftp-card__info-table', { [ 'is-placeholder' ]: loading } ) }
+				>
+					<tbody>
+						{ map( sftpData, ( data, title ) => (
+							<tr key={ title }>
+								<th>{ title }:</th>
+								<td>
+									<span>{ ! loading && data }</span>
+								</td>
+							</tr>
+						) ) }
+						<tr>
+							<th>{ translate( 'Password' ) }:</th>
+							<td>
+								<Button
+									onClick={ () => resetAtomicSFTPUserPassword( siteId ) }
+									disabled={ loading }
+								>
+									{ translate( 'Reset Password' ) }
+								</Button>
+							</td>
+						</tr>
+					</tbody>
+				</table>
+			</div>
+		</Card>
+	);
+};
+
+export default connect( state => {
+	const siteId = getSelectedSiteId( state );
+	const sftpDetails = requestAtomicSFTPDetails( siteId );
+	const username = get( sftpDetails, 'data.username', null );
```

Comment:
If we've established our parser and it only accepts an API response with the expected data and types then we can relax our type-check here too. note that we can perform semantic parsing and validation here too.

```js
fromApi: () => ( { data, error } ) => {
	if ( error ) {
		switch ( error.status ) {
			// expected errors explicitly typed out for searchability
			// and to show what types of errors we expect to handle
			case 'has_no_accounts':
				throw new Error( 'missing_account' );
			default:
				throw new Error( 'unknown_error' );
		}
	}

	return ( {
		username: data.username || null,
		password: data.password || null
	} )
}
```

Hopefully in seeing how our parsers can be constructed it demonstrates how we can fake the data in the component until the time comes when we're ready to hook it up. It gives you and @glendaviesnz a chance to discover the needs of the data structure and play with those before writing all the code so that if and when things need to change it's a simple change and not a bigger one.

Reply:
Thanks! That's also really helpful as a I get a handle on using our data layer.